dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/ApplicationProcessor.java

```diff
@@ -65,7 +77,12 @@
     
     @Reference
     private ClassLoadingService classLoadingService;
-
+    
+    // This needs to be an optional dynamic reference to accommodate the circular dependency
+    // It should almost always be present
+    @Reference(cardinality = OPTIONAL, policy = DYNAMIC, policyOption = GREEDY)
+    private volatile ApplicationRegistry appRegistry;
```

Comment:
What happens if ApplicationRegistry is null? If ApplicationRegistry and ApplicationProcessor depend on each other, why have them as separate components?

Reply:
The code that calls it checks for nullness first.

They're separate components since they're responsible for different things, the application processor handles processing a single application, the application registry keeps track of all the deployed applications. The link this direction is needed because we want to log a message if multiple modules are deployed but merging modules is disabled. The ApplicationRegistry is responsible for making sure this message is only emitted once.

Reply:
Moved the method which emits the message into its own component which can be reliably injected into both locations without causing a loop.